### DISCUSSION TITLE
About string definition

### DISCUSSION THREAD

**Discussion by Enes1313** [google/comprehensive-rust#2028] [General]
Hi, I am trying to translate this course into Turkish. Even if I understand what the two sentences about strings mean in English, I seem to be missing something (because my English is not very good).

Sentences: 
"String is an owned, heap-allocated buffer of UTF-8 bytes."
"String is the standard heap-allocated growable UTF-8 string buffer" 

What I understand is that:
"String is an owned" -> "String type has ownership over the contents of the string."
"String is an heap-allocated buffer of UTF-8 bytes." -> "String contains UTF-8 encoded bytes and String has heap allocated buffer"
"String is the standard heap-allocated growable UTF-8 string buffer"  -> "I understand a few things, but the sentence seems interesting."

What I don't understand is that:
"String is heap-allocated buffer of UTF-8 bytes." -> is String type is buffer?
"String is the standard heap-allocated growable UTF-8 string buffer" -> What is the "standad heap-allocated ..." ? I wonder if a comma is needed in some places?

Sorry my English,

**Comment by djmitche** [comment]
I think the English is using the term "buffer" here to mean "a block of
bytes in memory". To answer your question, "is String type is buffer", is
difficult -- String's type is String! But in some ways it works like a
buffer, in the sense that it owns a block of bytes in memory.

As for "standard heap-allocated", "standard" means it's used very widely in
Rust code. This word could probably apply to everything in `std`!
"heap-allocated" is differentiating `String` from `&'static str` (constant
strings embedded in the binary), which are not stored on the heap.

I hope that helps?

Dustin

**Reply by Enes1313** [reply]
Thank you you helped. I will feel a little more free for translation. I will translate the meanings of these sentences, not these sentences directly. I guess it won't be a problem.

I wanted to write here too, for the record:

"String is an owned, heap-allocated buffer of UTF-8 bytes." -> **"String type has ownership over the contents of the string and heap-allocated buffer of UTF-8 encoded bytes."**
"String is the standard heap-allocated growable UTF-8 string buffer" -> **"The String type represents a widely used heap-allocated growable UTF-8 encoded string buffer."**

I will translate as in the 2nd sentences, if there is no problem.

**Reply by mgeisler** [reply]
Hi @Enes1313, thanks for calling this out!

The English text is often hard to translate because it is using jargon or because it is condensed too much.

> I will translate the meanings of these sentences, not these sentences directly. I guess it won't be a problem.

That is absolutely fine! I think a good translation has to focus on the meaning to be understandable.

> "String is an owned, heap-allocated buffer of UTF-8 bytes." -> **"String type has ownership over the contents of the string and heap-allocated buffer of UTF-8 encoded bytes."**

Could we make this

> `String` has ownership over its heap-allocated buffer of UTF-8 encoded bytes.

instead in the English translation?

> "String is the standard heap-allocated growable UTF-8 string buffer" -> **"The String type represents a widely used heap-allocated growable UTF-8 encoded string buffer."**

Could you submit this as a change to the English text? I like it better than what we have currently :smile: 

I'm not a native English speaker (I'm originally from Denmark) and so the English you see in original is a mix between my style and many many fixed and improvements applied by tons of contributors. There are still small and big things to improve here, so your feedback is very valuable.

**Reply by Enes1313** [reply]
Hi @mgeisler , thank you for reply.

> That is absolutely fine! I think a good translation has to focus on the meaning to be understandable.

Yes, it is one of the rules I use for translation but these sentences seemed very strange to me 😄

> instead in the English translation?

Yes, it got better.

> Could you submit this as a change to the English text? I like it better than what we have currently 😄

Is this for two sentences or just second sentence?

> I'm not a native English speaker (I'm originally from Denmark) and so the English you see in original is a mix between my style and many many fixed and improvements applied by tons of contributors. There are still small and big things to improve here, so your feedback is very valuable.

I thought that the English used in the course was good terminological English. But after this discussion, I felt quite relieved.

**Reply by mgeisler** [reply]
> Is this for two sentences or just second sentence?

I think we should update both sentences: the first one to my suggestion and the second one to your suggestion :slightly_smiling_face: 

> I thought that the English used in the course was good terminological English. But after this discussion, I felt quite relieved.

That's good to hear! There are lots of opportunities for making the text better in small and big ways.

**Reply by Enes1313** [reply]
> I think we should update both sentences: the first one to my suggestion and the second one to your suggestion 🙂

Ok, I will open PR. Thanks for everything
